### PR TITLE
test: Migrate lto tests to snapbox

### PR DESCRIPTION
**Pull request by dieterplex**
Part of #14039.

### CODE REVIEW

**Comment by dieterplex** [comment]
Don't know why it fail twice.
```
thread 'global_cache_tracker::compatible_with_older_cargo' panicked at tests/testsuite/global_cache_tracker.rs:1987:10:

test failed running `/home/runner/work/cargo/cargo/target/debug/cargo check -Zgc`
error: process exited with code 101 (expected 0)
(...skip...)
thread 'main' panicked at src/cargo/core/compiler/fingerprint/mod.rs:1834:9:
assertion `left == right` failed
  left: "0b455c154b949b0d"
 right: "ce6d4698ea4438f2"
```
https://github.com/rust-lang/cargo/actions/runs/9833359314/job/27143461910#step:11:4050

**Comment by epage** [comment]
We're discussing that failure at https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Fingerprint.20assert.20in.20tests

File: tests/testsuite/lto.rs [hunk 1]
```diff
@@ -192,33 +231,47 @@ fn complicated() {
...
-        .with_stderr_contains(
-            "[..]`rustc[..]--crate-name dep_proc_macro [..]-C embed-bitcode=no[..]`",
-        )
-        .with_stderr_contains(
-            "[..]`rustc[..]--crate-name foo_bin [..]--crate-type bin[..]-C lto[..]`",
-        )
-        .with_stderr_contains(
-            "[..]`rustc[..]--crate-name test [..]--crate-type cdylib[..]-C lto[..]`",
+        // - normal deps and their transitive dependencies do not need object
+        //   code, so they should have linker-plugin-lto specified
```
Comment (epage) [inline tests/testsuite/lto.rs:235]: Let's revert this one.  I don't trust `does_not_contain` much but pairing them with easily read `contains` that make it easy to see when one needs updating, helps.

**Comment by dieterplex** [comment]
Now it breaks with
```

thread 'main' has overflowed its stack
error: could not document `foo`

Caused by:
  process didn't exit successfully: `rustdoc --edition=2015 --crate-type lib --crate-name foo src\lib.rs -o D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\doc --check-cfg cfg(docsrs) --check-cfg "cfg(feature, values())" --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=10d924ec6a37bdaa -L dependency=D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps --extern a=D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps\liba-afbe6803dac1e694.rmeta --crate-version 0.0.1` (exit code: 0xc00000fd, STATUS_STACK_OVERFLOW)

failures:
    doc::doc_cap_lints
```

should it rerun CI?

**Comment by weihanglo** [comment]
Attached full log here for futurue reference.

```
---- doc::doc_cap_lints stdout ----
running `D:\a\cargo\cargo\target\debug\cargo.exe doc`
running `D:\a\cargo\cargo\target\debug\cargo.exe doc -vv`
thread 'doc::doc_cap_lints' panicked at tests\testsuite\doc.rs:1872:10:

test failed running `D:\a\cargo\cargo\target\debug\cargo.exe doc -vv`
error: process exited with code 101 (expected 0)
--- stdout

--- stderr
    Checking a v0.5.0 (file:///D:/a/cargo/cargo/target/tmp/cit/t1268/a#6781128e)
 Documenting a v0.5.0 (file:///D:/a/cargo/cargo/target/tmp/cit/t1268/a#6781128e)
     Running `set CARGO=\\?\D:\a\cargo\cargo\target\debug\cargo.exe&& set CARGO_CRATE_NAME=a&& set CARGO_MANIFEST_DIR=D:\a\cargo\cargo\target\tmp\cit\t1268\home\.cargo\git\checkouts\a-8baaf4225d65e0c6\6781128&& set CARGO_PKG_AUTHORS=wycats@example.com&& set CARGO_PKG_DESCRIPTION=""&& set CARGO_PKG_HOMEPAGE=""&& set CARGO_PKG_LICENSE=""&& set CARGO_PKG_LICENSE_FILE=""&& set CARGO_PKG_NAME=a&& set CARGO_PKG_README=""&& set CARGO_PKG_REPOSITORY=""&& set CARGO_PKG_RUST_VERSION=""&& set CARGO_PKG_VERSION=0.5.0&& set CARGO_PKG_VERSION_MAJOR=0&& set CARGO_PKG_VERSION_MINOR=5&& set CARGO_PKG_VERSION_PATCH=0&& set CARGO_PKG_VERSION_PRE=""&& set PATH="D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps;C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-gnu\bin;D:\a\cargo\cargo\target\debug\build\curl-sys-3cdfc50fd064581e\out\build;D:\a\cargo\cargo\target\debug\build\libgit2-sys-fe666c639a2e07ee\out\build;D:\a\cargo\cargo\target\debug\build\libnghttp2-sys-13fcb8b547fbe868\out\i\lib;D:\a\cargo\cargo\target\debug\build\libsqlite3-sys-37c5e72bff0a7ce3\out;D:\a\cargo\cargo\target\debug\build\libssh2-sys-8856f938628c929f\out\build;D:\a\cargo\cargo\target\debug\build\libz-sys-0d64be973d275c44\out\lib;D:\a\cargo\cargo\target\debug\build\libz-sys-896f62a45e9e91c5\out\lib;D:\a\cargo\cargo\target\debug\deps;D:\a\cargo\cargo\target\debug;C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\lib;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Users\runneradmin\bin;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\Program Files (x86)\NSIS\;C:\tools\zstd;C:\Program Files\Mercurial\;C:\hostedtoolcache\windows\stack\2.15.7\x64;C:\cabal\bin;C:\\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files\R\R-4.4.1\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver\;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.21.11\x64\bin;C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.13\x64;C:\hostedtoolcache\windows\Ruby\3.0.7\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.412-8\x64\bin;C:\Program Files\ImageMagick-7.1.1-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\docker-compose;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\PowerShell\7\;C:\Program Files\Microsoft\Web Platform Installer\;C:\Program Files\TortoiseSVN\bin;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files\Microsoft SQL Server\150\Tools\Binn\;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files (x86)\WiX Toolset v3.14\bin;C:\Program Files\Microsoft SQL Server\130\DTS\Binn\;C:\Program Files\Microsoft SQL Server\140\DTS\Binn\;C:\Program Files\Microsoft SQL Server\150\DTS\Binn\;C:\Program Files\Microsoft SQL Server\160\DTS\Binn\;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.7\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs\;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI\;c:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2\;C:\Program Files\Amazon\SessionManagerPlugin\bin\;C:\Program Files\Amazon\AWSSAMCLI\bin\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps"&& rustc --crate-name a --edition=2015 D:\a\cargo\cargo\target\tmp\cit\t1268\home\.cargo\git\checkouts\a-8baaf4225d65e0c6\6781128\src\lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata -C embed-bitcode=no -C debuginfo=2 --check-cfg cfg(docsrs) --check-cfg "cfg(feature, values())" -C metadata=afbe6803dac1e694 -C extra-filename=-afbe6803dac1e694 --out-dir D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps -L dependency=D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps --cap-lints warn`
     Running `rustdoc --edition=2015 --crate-type lib --crate-name a D:\a\cargo\cargo\target\tmp\cit\t1268\home\.cargo\git\checkouts\a-8baaf4225d65e0c6\6781128\src\lib.rs --cap-lints warn -o D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\doc --check-cfg cfg(docsrs) --check-cfg "cfg(feature, values())" --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=afbe6803dac1e694 -L dependency=D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps --crate-version 0.5.0`
warning: lint `broken_intra_doc_links` has been renamed to `rustdoc::broken_intra_doc_links`
 --> D:\a\cargo\cargo\target\tmp\cit\t1268\home\.cargo\git\checkouts\a-8baaf4225d65e0c6\6781128\src\lib.rs:2:9
  |
2 | #![deny(broken_intra_doc_links)]
  |         ^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `rustdoc::broken_intra_doc_links`
  |
  = note: `#[warn(renamed_and_removed_lints)]` on by default

warning: unresolved link to `bad_link`
 --> D:\a\cargo\cargo\target\tmp\cit\t1268\home\.cargo\git\checkouts\a-8baaf4225d65e0c6\6781128\src\lib.rs:4:6
  |
4 | /// [bad_link]
  |      ^^^^^^^^ no item named `bad_link` in scope
  |
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
note: the lint level is defined here
 --> D:\a\cargo\cargo\target\tmp\cit\t1268\home\.cargo\git\checkouts\a-8baaf4225d65e0c6\6781128\src\lib.rs:2:9
  |
2 | #![deny(broken_intra_doc_links)]
  |         ^^^^^^^^^^^^^^^^^^^^^^

warning: `a` (lib doc) generated 2 warnings
 Documenting foo v0.0.1 (D:\a\cargo\cargo\target\tmp\cit\t1268\foo)
     Running `rustdoc --edition=2015 --crate-type lib --crate-name foo src\lib.rs -o D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\doc --check-cfg cfg(docsrs) --check-cfg "cfg(feature, values())" --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=10d924ec6a37bdaa -L dependency=D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps --extern a=D:\a\cargo\cargo\target\tmp\cit\t1268\foo\target\debug\deps\liba-afbe6803dac1e694.rmeta --crate-version 0.0.1`

thread 'main' has overflowed its stack
error: could not document `foo`
```

It doesn't make sense that overflow only happened on Windows GNU.

@bors try

File: tests/testsuite/lto.rs [hunk 2]
```diff
@@ -517,45 +579,48 @@ fn cdylib_and_rlib() {
...
-[RUNNING] `rustc --crate-name foo [..]-C lto [..]--test[..]
-[RUNNING] `rustc --crate-name a [..]-C lto [..]--test[..]
-[FINISHED] [..]
-[RUNNING] [..]
-[RUNNING] [..]
-",
+[FRESH] bar v0.0.0 ([ROOT]/foo/bar)
+[COMPILING] foo v0.0.0 ([ROOT]/foo)
+[RUNNING] `rustc --crate-name foo [..] -C lto --test [..]`
+[RUNNING] `rustc --crate-name a [..] -C lto --test [..]`
```
Comment (weihanglo) [inline tests/testsuite/lto.rs:589]: Could we put `[..]` back between `-C lto` and `--test`? We don't want snapshots to depend on the order of these flags.

Also, the missing space around `[..]` is important for not overly-matching the order

```suggestion
[RUNNING] `rustc --crate-name foo [..]-C lto [..]--test [..]`
[RUNNING] `rustc --crate-name a [..]-C lto [..]--test [..]`
```

(ditto elsewhere throughout this PR)

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Thank you for your contribution!

@bors r+

**Pull request merged**